atlasdb-commons/src/test/java/com/palantir/common/compression/LZ4CompressionTests.java

```diff
@@ -0,0 +1,155 @@
+        initializeStreams(uncompressedData);
+        int value = decompressingStream.read();
+
+        assertEquals(uncompressedData[0], value);
+        assertStreamIsEmpty(decompressingStream);
+    }
+
+    @Test
+    public void testSingleBlock() throws IOException {
+        byte[] uncompressedData = new byte[BLOCK_SIZE];
+        fillWithCompressibleData(uncompressedData);
+        initializeStreams(uncompressedData);
+        verifyStreamContents(uncompressedData);
+    }
+
+    @Test
+    public void testSingleBlock_incompressible() throws IOException {
+        byte[] uncompressedData = new byte[BLOCK_SIZE];
+        fillWithIncompressibleData(uncompressedData);
+        initializeStreams(uncompressedData);
+        verifyStreamContents(uncompressedData);
+    }
+
+    @Test
+    public void testMultiBlock_compressible() throws IOException {
+        byte[] uncompressedData = new byte[16 * BLOCK_SIZE];
+        fillWithCompressibleData(uncompressedData);
+        initializeStreams(uncompressedData);
+        verifyStreamContents(uncompressedData);
+    }
+
+    @Test
+    public void testMultiBlock_incompressible() throws IOException {
+        byte[] uncompressedData = new byte[16 * BLOCK_SIZE];
+        fillWithIncompressibleData(uncompressedData);
+        initializeStreams(uncompressedData);
+        verifyStreamContents(uncompressedData);
+    }
+
+    @Test
+    public void testMultiBlock_singleByteReads() throws IOException {
+        byte[] uncompressedData = new byte[16 * BLOCK_SIZE];
+        fillWithIncompressibleData(uncompressedData);
+        initializeStreams(uncompressedData);
+
+        for (int i = 0; i < uncompressedData.length; ++i) {
+            int value = decompressingStream.read();
+            assertEquals(Byte.toUnsignedInt(uncompressedData[i]), value);
+        }
+        assertStreamIsEmpty(decompressingStream);
+    }
+
+    @Test
+    public void testMultiBlock_readPastEnd() throws IOException {
+        byte[] uncompressedData = new byte[16 * BLOCK_SIZE];
+        fillWithCompressibleData(uncompressedData);
+        initializeStreams(uncompressedData);
+
+        byte[] decompressedData = new byte[17 * BLOCK_SIZE];
+        int bytesRead = ByteStreams.read(decompressingStream, decompressedData, 0, decompressedData.length);
+        assertEquals(uncompressedData.length, bytesRead);
+        assertArrayEquals(uncompressedData, Arrays.copyOf(decompressedData, bytesRead));
+    }
+
+    private void initializeStreams(byte[] uncompressedData) throws IOException {
+        uncompressedStream = new ByteArrayInputStream(uncompressedData);
+        compressingStream = new LZ4CompressingInputStream(uncompressedStream);
+        decompressingStream = new LZ4BlockInputStream(compressingStream);
+    }
+
+    private void fillWithCompressibleData(byte[] data) {
+        Arrays.fill(data, SINGLE_VALUE);
+    }
+
+    private void fillWithIncompressibleData(byte[] data) {
+        new Random(0).nextBytes(data);
```

Comment:
I verified that it was while writing the test - the output data from the compressing stream was larger than the input data.